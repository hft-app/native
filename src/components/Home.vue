<template>
  <div class="wrapper">
    <div class="dashboard">
      <router-link v-if="lectures.length > 0" class="widget" to="timetable">
        <h5 class="container">{{ $t('page.home.nextLectures') }}</h5>
        <div v-for="lecture in lectures" v-hash-color="lecture.title" class="lecture">
          <div class="container">
            <div v-if="lecture.start < now" class="start">  {{ $t('page.home.currently') }}</div>
            <div v-else-if="Math.floor(lecture.start/864E5) === Math.floor(now/864E5)" class="start">
              {{ $d(lecture.start, 'time') }}
            </div>
            <div v-else-if="Math.floor(lecture.start/864E5) === Math.floor(now/864E5) + 1" class="start">
              {{ $t('page.home.tomorrow') }}
            </div>
            <div v-else class="start">{{ $d(lecture.start, 'day') }}</div>

            <div class="title">{{ lecture.title }}</div>
            <div class="data">
              <div v-if="lecture.professor" class="info">
                <fa-icon icon="user-alt" />
                {{ lecture.professor }}
              </div>
              <div v-if="lecture.room" class="info">
                <fa-icon icon="map-marker-alt" />
                {{ lecture.room }}
              </div>
            </div>
          </div>
        </div>
      </router-link>
      <div v-if="events.length > 0" class="widget">
        <h5 class="container">{{ $t('page.home.nextEvents') }}</h5>
        <div class="container">
          <EventItem v-for="event in events" :key="event" :event="event" />
        </div>
      </div>
    </div>
    <div class="note">
      <div class="container">
        <p>{{ $t('page.home.lastRefresh') }}</p>
        <p>{{ $tc('page.home.lastRefreshDuration', Math.floor(lastRefresh / 36E5)) }}</p>
      </div>
    </div>
  </div>
</template>
<script>
  import EventItem from './EventItem.vue';

  export default {
    components: {
      EventItem
    },
    data() {
      return {
        now: Date.now()
      }
    },
    computed: {
      lectures() {
        return this.$store.state.lsf.lectures
          .filter(lecture => lecture.end > this.now)
          .slice(0, 2)
      },
      events() {
        return this.$store.state.hft.events
          .filter(event => event.endDate &&
            event.endDate + 864E5 >= this.now || event.startDate >= this.now)
          .slice(0, 3)
      },
      lastRefresh() {
        return Math.max(this.now - this.$store.state.lastRefresh, 0)
      }
    }
  }
</script>
<style scoped lang="scss">
  @import "../colors";

  .widget {
    display: block;

    &.widget {
      margin-top: 2em
    }
  }

  h5 {
    padding-bottom: 0.8em;
    color: $secondary;
    font-weight: 500;
    font-size: 0.9em;
  }

  .lecture {
    padding: 0.8em 0;
    color: #FFF;

    &.lecture {
      margin-top: 1px;
    }

    .container {
      display: flex;
      align-items: center;
      justify-content: space-between;
    }

    .start {
      flex-shrink: 0;
      padding: 4px 6px 3px;
      border-radius: 100px;
      font-size: 0.6em;
      background-color: $background;
      color: $text;
    }

    .title {
      flex-grow: 1;
      padding: 0 0.8em;
      font-weight: 500;
    }

    .data {
      text-align: right;
      font-weight: 300;
      font-size: 0.8em;
      flex-shrink: 0;

      .info + .info {
        margin-top: 0.5em;
      }
    }
  }

  .note {
    margin-top: 30px;
  }
</style>
