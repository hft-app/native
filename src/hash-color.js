import Vue from 'vue';

Vue.directive('hash-color', {
  bind(el, {value}) {
    const colors = ['green', 'blue', 'orange', 'red', 'yellow', 'teal', 'purple', 'pink'];
    let hash = 0;
    for(let i = 0; i < value.length; i++) {
      hash += value.charCodeAt(i);
    }
    hash %= colors.length;
    el.classList.add(colors[hash]);
  }
});
