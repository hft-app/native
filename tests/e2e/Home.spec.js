import {config, createLocalVue, shallowMount} from '@vue/test-utils'
import Home from 'components/Home.vue';
import Vuex from 'vuex';
import 'hash-color'

config.mocks['$d'] = date => date.valueOf();
config.mocks['$t'] = str => str;
config.mocks['$tc'] = (str, n) => str + n;

describe('Home', () => {
  test('renders correctly', () => {
    const localVue = createLocalVue();
    localVue.use(Vuex);

    const store = new Vuex.Store({
      state: {
        lastRefresh: new Date('Mon Feb 19 2020 09:30:00 GMT+0100').valueOf()
      },
      modules: {
        lsf: {
          state: {
            lectures: [{
              title: 'Algorithmische Geometrie',
              room: '2/433',
              start: new Date('Mon Feb 19 2020 08:00:00 GMT+0100'),
              end: new Date('Mon Feb 19 2020 09:30:00 GMT+0100'),
              professor: 'Müller'
            }, {
              title: 'Geo-Visualisierung',
              professor: 'Müller',
              start: new Date('Mon Feb 19 2020 10:30:00 GMT+0100'),
              end: new Date('Mon Feb 19 2020 12:00:00 GMT+0100')
            }, {
              title: 'Geschäftsprozessmanagement',
              start: new Date('Mon Feb 20 2020 08:00:00 GMT+0100'),
              end: new Date('Mon Feb 20 2020 09:30:00 GMT+0100')
            }, {
              title: 'Infoveranstaltung Informatik Bachelor-Thesis und Ausland',
              room: '5/U26',
              start: new Date('Mon Feb 21 2020 08:00:00 GMT+0100'),
              end: new Date('Mon Feb 21 2020 09:30:00 GMT+0100')
            }]
          }
        },
        hft: {
          state: {
            events: [{
              title: 'Stallfest',
              description: '@ Block4',
              startDate: new Date('Mon Feb 20 2020 03:00:00 GMT+0100')
            }]
          }
        }
      }
    });

    const wrapper = shallowMount(Home, {
      stubs: ['fa-icon', 'router-link'],
      localVue,
      store,
      data() {
        return {now: new Date('Mon Feb 19 2020 11:00:00 GMT+0100')}
      }
    });

    expect(wrapper.element).toMatchSnapshot()
  });
});
