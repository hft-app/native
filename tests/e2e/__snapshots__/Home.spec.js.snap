// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders correctly 1`] = `
<div
  class="wrapper"
>
  <div
    class="dashboard"
  >
    <router-link-stub
      class="widget"
      to="timetable"
    >
      <h5
        class="container"
      >
        page.home.nextLectures
      </h5>
       
      <div
        class="lecture orange"
      >
        <div
          class="container"
        >
          <div
            class="start"
          >
            Jetzt
          </div>
           
          <div
            class="title"
          >
            Geo-Visualisierung
          </div>
           
          <div
            class="data"
          >
            <div
              class="info"
            >
              <fa-icon-stub
                icon="user-alt"
              />
              
              Müller
            
            </div>
             
            <!---->
          </div>
        </div>
      </div>
      <div
        class="lecture purple"
      >
        <div
          class="container"
        >
          <div
            class="start"
          >
            
            Morgen
          
          </div>
           
          <div
            class="title"
          >
            Geschäftsprozessmanagement
          </div>
           
          <div
            class="data"
          >
            <!---->
             
            <!---->
          </div>
        </div>
      </div>
    </router-link-stub>
     
    <div
      class="widget"
    >
      <h5
        class="container"
      >
        page.home.nextEvents
      </h5>
       
      <div
        class="container"
      >
        <eventitem-stub
          event="[object Object]"
        />
      </div>
    </div>
  </div>
   
  <div
    class="note"
  >
    <div
      class="container"
    >
      <p>
        page.home.lastRefresh
      </p>
       
      <p>
        page.home.lastRefreshDuration1
      </p>
    </div>
  </div>
</div>
`;
